# 기계 학습 실기 평가 -임해찬- 2020.06.26

# 1 회귀식 및 RMSE구하기

In [257]:
from sklearn.datasets import load_iris
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from sklearn.linear_model import LinearRegression
from sklearn.metrics import accuracy_score

from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression

import numpy as np
import warnings
warnings.filterwarnings('ignore')
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.preprocessing import scale
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_samples, silhouette_score

import re
from konlpy.tag import Okt

iris = load_iris()
iris_data = iris.data
iris_label = iris.target

iris_df = pd.DataFrame(data=iris_data, columns=iris.feature_names)
iris_df['label'] = iris.target
iris_df

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
..                 ...               ...                ...               ...   
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

     label  
0        0  
1        0  
2        0  
3        0  
4        0  
..     ...  
145      2  
146      2  
147      2  
148      2  
149      2  

[150 rows x 5 columns]

In [258]:
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   label              150 non-null    int32  
dtypes: float64(4), int32(1)
memory usage: 5.4 KB


In [259]:
iris_df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)       label  
count        150.000000  150.000000  
mean           1.199333    1.000000  
std            0.762238    0.819232  
min            0.100000    0.000000  
25%            0.300000    0.000000  
50%            1.300000    1.000000  
75%            1.800000    2.000000  
max            2.500000    2.000000

In [260]:
X = iris_df['sepal width (cm)']
Y = iris_df.drop(['sepal width (cm)'], axis = 1)

In [261]:
X_train, X_test, y_train, y_test = train_test_split(X, Y, 
                                                    test_size=0.2, random_state=2020)

In [262]:
sim_lr = LinearRegression()

In [263]:
X_train.values.reshape(-1,1)

array([[3. ],
       [3.5],
       [3.4],
       [3.5],
       [3.2],
       [3.1],
       [3.3],
       [2.4],
       [3.2],
       [2.5],
       [3.5],
       [3.3],
       [3. ],
       [3.8],
       [2.9],
       [3.2],
       [3.8],
       [2.8],
       [3.1],
       [2.4],
       [3.8],
       [2.5],
       [2.5],
       [3.5],
       [3.3],
       [2.6],
       [3.1],
       [2.9],
       [4.4],
       [2.3],
       [4.1],
       [3. ],
       [3.5],
       [2.7],
       [2.8],
       [3.2],
       [3.4],
       [3.8],
       [3.4],
       [2.9],
       [2.8],
       [3.9],
       [3.6],
       [2.9],
       [3.2],
       [2.8],
       [3.6],
       [3. ],
       [2.8],
       [3.4],
       [3. ],
       [2. ],
       [3.4],
       [3.8],
       [2.8],
       [3. ],
       [3.1],
       [3.2],
       [2.9],
       [2.7],
       [3.7],
       [2.5],
       [2.7],
       [3.9],
       [3. ],
       [3. ],
       [4. ],
       [3. ],
       [2.7],
       [2.5],
       [2.9],
      

In [264]:
sim_lr.fit(X_train.values.reshape(-1,1), y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [265]:
y_pred = sim_lr.predict(X_test.values.reshape(-1,1))

In [266]:
r2_score(y_test, y_pred)

0.07356351338272699

In [267]:
sim_lr.coef_, sim_lr.intercept_

(array([[-0.27199255],
        [-1.7759912 ],
        [-0.65345068],
        [-0.76164685]]),
 array([6.72909703, 9.25559951, 3.21422869, 3.34968053]))

### 도출한 회귀식
### sepal_length = -0.272 * sepal width + 6.729
### petal length = -1.776 * sepal width + 9.256
### petal width = -0.653 * sepal width + 3.214
### variety = -0.762 * sepal width + 3.350

In [268]:
from sklearn.model_selection import cross_val_score
from sklearn.metrics import mean_squared_error

mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)

print('MSE : {0:.3f}, RMSE : {1:.3F}'.format(mse , rmse))

MSE : 0.971, RMSE : 0.986


# 2 폐암환자 생존여부 예측 및 각각의 정확도 구하기

In [269]:
Surgery_df = pd.read_csv('./data/ThoraricSurgery.csv')
Surgery_df

293  1   3.8   2.8  0  0.1  0.2  0.3  0.4  0.5  12  0.6  0.7  0.8  1.1  \
0      1  2  2.88  2.16  1    0    0    0    1    1  14    0    0    0    1   
1      8  2  3.19  2.50  1    0    0    0    1    0  11    0    0    1    1   
2     14  2  3.98  3.06  2    0    0    0    1    1  14    0    0    0    1   
3     17  2  2.21  1.88  0    0    1    0    0    0  12    0    0    0    1   
4     18  2  2.96  1.67  0    0    0    0    0    0  12    0    0    0    1   
..   ... ..   ...   ... ..  ...  ...  ...  ...  ...  ..  ...  ...  ...  ...   
464   98  6  3.04  2.40  2    0    0    0    1    0  11    0    0    0    1   
465  369  6  3.88  2.72  1    0    0    0    1    0  12    0    0    0    1   
466  406  6  5.36  3.96  1    0    0    0    1    0  12    0    0    0    0   
467   25  8  4.32  3.20  0    0    0    0    0    0  11    0    0    0    0   
468  447  8  5.20  4.10  0    0    0    0    0    0  12    0    0    0    0   

     0.9  62  0.10  
0      0  60     0  
1      0  66     1  
2      0  80     1  
3      0  56     0  
4      0  61     0  
..   ...  ..   ...  
464    0  76     0  
465    0  77     0  
466    0  62     0  
467    0  58     1  
468    0  49     0  

[469 rows x 18 columns]

In [270]:
y_Surgery_df = titanic_df['0.10']
X_Surgery_df= titanic_df.drop('0.10', axis=1)

X_train, X_test, y_train, y_test=train_test_split(X_Surgery_df, y_Surgery_df, \
                                                  test_size=0.2, random_state=2020)

In [271]:
X_Surgery_df

293  1   3.8   2.8  0  0.1  0.2  0.3  0.4  0.5  12  0.6  0.7  0.8  1.1  \
0      1  2  2.88  2.16  1    0    0    0    1    1  14    0    0    0    1   
1      8  2  3.19  2.50  1    0    0    0    1    0  11    0    0    1    1   
2     14  2  3.98  3.06  2    0    0    0    1    1  14    0    0    0    1   
3     17  2  2.21  1.88  0    0    1    0    0    0  12    0    0    0    1   
4     18  2  2.96  1.67  0    0    0    0    0    0  12    0    0    0    1   
..   ... ..   ...   ... ..  ...  ...  ...  ...  ...  ..  ...  ...  ...  ...   
464   98  6  3.04  2.40  2    0    0    0    1    0  11    0    0    0    1   
465  369  6  3.88  2.72  1    0    0    0    1    0  12    0    0    0    1   
466  406  6  5.36  3.96  1    0    0    0    1    0  12    0    0    0    0   
467   25  8  4.32  3.20  0    0    0    0    0    0  11    0    0    0    0   
468  447  8  5.20  4.10  0    0    0    0    0    0  12    0    0    0    0   

     0.9  62  
0      0  60  
1      0  66  
2      0  80  
3      0  56  
4      0  61  
..   ...  ..  
464    0  76  
465    0  77  
466    0  62  
467    0  58  
468    0  49  

[469 rows x 17 columns]

In [272]:
svc = SVC()
svc.fit(X_train, y_train)
y_pred_svc = svc.predict(X_test)
accuracy_score(y_test, y_pred_svc)

0.8936170212765957

In [273]:
dtc = DecisionTreeClassifier(random_state=2020)
dtc.fit(X_train ,y_train)
y_pred_dtc = dtc.predict(X_test)
accuracy_score(y_test, y_pred_dtc)

0.7872340425531915

In [274]:
lgt = LogisticRegression()
lgt.fit(X_train, y_train)
y_pred_lgt = lgt.predict(X_test)
accuracy_score(y_test, y_pred_lgt)

0.8936170212765957

In [275]:
print('SVM 정확도 : %.3f' %accuracy_score(y_test, y_pred_svc))
print('결정 트리 정확도 : %.3f' %accuracy_score(y_test, y_pred_dtc))
print('로지스틱 회귀 정확도 : %.3f' %accuracy_score(y_test, y_pred_lgt))

SVM 정확도 : 0.894
결정 트리 정확도 : 0.787
로지스틱 회귀 정확도 : 0.894


# 3 Iris 데이터 2차원으로 차원 축소 후 K = 2, 3, 4에 대해 군집화 해보고 성능 평가하기

In [276]:
columns = ['sepal_length','sepal_width','petal_length','petal_width']
irisDF = pd.DataFrame(iris.data, columns=columns)
irisDF['target']=iris.target
irisDF.head(3)

sepal_length  sepal_width  petal_length  petal_width  target
0           5.1          3.5           1.4          0.2       0
1           4.9          3.0           1.4          0.2       0
2           4.7          3.2           1.3          0.2       0

In [277]:
iris_scaled = StandardScaler().fit_transform(irisDF.iloc[:,:-1])
pca = PCA(n_components=2)

pca.fit(iris_scaled)
iris_pca = pca.transform(iris_scaled)
print(iris_pca.shape)

(150, 2)


In [278]:
pca_columns=['pca_component_1','pca_component_2']
irisDF_pca = pd.DataFrame(iris_pca, columns=pca_columns)
irisDF_pca['target']=iris.target
irisDF_pca

pca_component_1  pca_component_2  target
0          -2.264703         0.480027       0
1          -2.080961        -0.674134       0
2          -2.364229        -0.341908       0
3          -2.299384        -0.597395       0
4          -2.389842         0.646835       0
..               ...              ...     ...
145         1.870503         0.386966       2
146         1.564580        -0.896687       2
147         1.521170         0.269069       2
148         1.372788         1.011254       2
149         0.960656        -0.024332       2

[150 rows x 3 columns]

In [279]:
kmeans = KMeans(n_clusters=5, init='k-means++', max_iter=300,
                random_state=2020).fit(irisDF_pca)

irisDF_pca['cluster'] = kmeans.labels_

In [280]:
score_samples = silhouette_samples(iris_pca.data, irisDF_pca['cluster'])
print('silhouette_samples( ) return 값의 shape', score_samples.shape)

irisDF_pca['silhouette_coeff'] = score_samples

average_score = silhouette_score(iris_pca.data, irisDF_pca['cluster'])
print('붓꽃 데이터셋 Silhouette Analysis Score:{0:.3f}'.format(average_score))

irisDF_pca

silhouette_samples( ) return 값의 shape (150,)
붓꽃 데이터셋 Silhouette Analysis Score:0.404


pca_component_1  pca_component_2  target  cluster  silhouette_coeff
0          -2.264703         0.480027       0        0          0.240866
1          -2.080961        -0.674134       0        3          0.667667
2          -2.364229        -0.341908       0        3          0.643183
3          -2.299384        -0.597395       0        3          0.679458
4          -2.389842         0.646835       0        0          0.407425
..               ...              ...     ...      ...               ...
145         1.870503         0.386966       2        4          0.428657
146         1.564580        -0.896687       2        2          0.244488
147         1.521170         0.269069       2        4          0.059737
148         1.372788         1.011254       2        4          0.279379
149         0.960656        -0.024332       2        2          0.570615

[150 rows x 5 columns]

In [281]:
irisDF_pca.groupby('cluster')['silhouette_coeff'].mean()

cluster
0    0.422742
1    0.437806
2    0.354445
3    0.486002
4    0.369907
Name: silhouette_coeff, dtype: float64

### K = 2 일때 0.354445
### K = 3 일때 0.486002
### K = 4 일때 0.369907

# 4 미드웨이 영화 평점에 따라 감성분석하기

In [282]:
review_df = pd.read_csv('./data/midway.tsv', 
                        header=0, sep="\t", quoting=3)
review_df

Unnamed: 0  평점                일시  \
0              0   9  2019.12.31 09:48   
1              1  10  2019.12.31 10:41   
2              2  10  2019.12.31 09:38   
3              3  10  2019.12.31 09:14   
4              4  10  2019.12.31 19:58   
...          ...  ..               ...   
4820        4820   1  2019.12.31 09:11   
4821        4821   2  2020.01.05 20:41   
4822        4822   2  2020.01.01 23:40   
4823        4823   4  2019.12.31 15:37   
4824        4824   1  2020.01.16 23:51   

                                                    감상평  
0                            미드웨이가 재밌으면 추천 백두산이 재밌으면 비추  
1                             저 해전이 있었기에 우리나라 광복도 가능 했음  
2                          백두산 상영관 대폭줄이고 미드웨이 상영관 대폭늘려라  
3         방금 개봉했는데 1점 준애는 뭐냐 ㅋㅋ 일본놈이냐? 이제 광고돌고 있을 시간이다.  
4     교과서에는 미드웨이 해전에서 미국이 승리했다고 한 줄로 서술되어 있지만 단순히 한 ...  
...                                                 ...  
4820  대만에서 10월31일 개봉했었는데.성공못했음. 미국에서도 성적이 별로구..그이유가 ...  
4821  아니 이게 재밌다는 사람은 뭐야?;그냥 국뽕->미국뽕 이 차이밖에 없는데 이 영화가...  
4822                  시간이 아까워요. 엉성한 그래서 감동도 교훈도 없는...ㅉㅉ  
4823                   진주만보다 낫다고해서봤는데대실망...천조국,왜놈들 국뽕영화  
4824  우리는 왜 이런 영화 못만드냐? 싸구려 좌파 찌라시 백x산이나 대놓고 친북사상 반일...  

[4825 rows x 4 columns]

In [283]:
review_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4825 entries, 0 to 4824
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  4825 non-null   int64 
 1   평점          4825 non-null   int64 
 2   일시          4825 non-null   object
 3   감상평         4316 non-null   object
dtypes: int64(2), object(2)
memory usage: 150.9+ KB


In [284]:
print(review_df['감상평'])

0                              미드웨이가 재밌으면 추천 백두산이 재밌으면 비추
1                               저 해전이 있었기에 우리나라 광복도 가능 했음
2                            백두산 상영관 대폭줄이고 미드웨이 상영관 대폭늘려라
3           방금 개봉했는데 1점 준애는 뭐냐 ㅋㅋ 일본놈이냐? 이제 광고돌고 있을 시간이다.
4       교과서에는 미드웨이 해전에서 미국이 승리했다고 한 줄로 서술되어 있지만 단순히 한 ...
                              ...                        
4820    대만에서 10월31일 개봉했었는데.성공못했음. 미국에서도 성적이 별로구..그이유가 ...
4821    아니 이게 재밌다는 사람은 뭐야?;그냥 국뽕->미국뽕 이 차이밖에 없는데 이 영화가...
4822                    시간이 아까워요. 엉성한 그래서 감동도 교훈도 없는...ㅉㅉ
4823                     진주만보다 낫다고해서봤는데대실망...천조국,왜놈들 국뽕영화
4824    우리는 왜 이런 영화 못만드냐? 싸구려 좌파 찌라시 백x산이나 대놓고 친북사상 반일...
Name: 감상평, Length: 4825, dtype: object


In [285]:
from sklearn.model_selection import train_test_split

class_df = review_df['평점']
feature_df = review_df.drop(['Unnamed: 0','평점'], axis=1, inplace=False)

X_train, X_test, y_train, y_test = \
    train_test_split(feature_df, class_df, test_size=0.2, random_state=2020)

X_train.shape, X_test.shape

((3860, 2), (965, 2))

# 마지막 문제는 어렵네요 수고하셨습니다.